bindings/js/api.js

```diff
@@ -0,0 +1,285 @@
+
+  // The WebAssembly modules have to be initialized before calling any function.
+  // This checks if it has been done already, and if not does it.
+  const checkIfInitialized = function() {
+    if (isInitialized === false) {
+      return Promise.all(shell.my_modules.map(m => {
+        var source = fs.readFileSync(path.resolve( __dirname, './' + m + ".wasm"));
+        return new Uint8Array(source)
+      })).then(bufs => {
+        return loader.link(my_imports, bufs.map((b, i) => ({
+          buf: b,
+          name: shell.my_modules[i]
+        })))
+      }).then(scope => {
+        Module = scope;
+        isInitialized = true;
+      })
+    } else {
+      return Promise.resolve()
+    }
+  }
+
+  /*
+  Inside WebAssembly, the functions only take pointers to memory and integers.
+  However, we want to expose the functions of the wasm module with a nice Javascript
+  API that manipulates ArrayBuffers.
+
+  In order to do that, we have to describe the Javascript prototype of each function
+  that we expose. The functions can take and return multiple objects that can be
+  buffers, integers or booleans. The buffers can either have a fixed length (and
+  in that case we check dynamically whether they have the correct length), or
+  have a variable length (and we have to pass that length as an additional
+  parameter to WebAssembly).
+
+  In order to match the Javascript API with the actual calls to WebAssembly functions,
+  we have to describe the correspondance between the two in the `api.json` file.
+
+  The scheme of the JSON file is the following :
+  - `module`, this module name will be shown in the JS API
+    - `function`, this function name will be shown in the JS API
+      - `module`, the name of the WebAssembly file where to find the function
+      - 'name', the name of the WebAssembly export corresponding to the function
+      - 'args', the list of the WebAssembly arguments expected by the function
+        - 'name', the name of the argument which will be shown in the JS Doc
+        - 'kind', either `input` or `output` of the function
+        - 'type', either 'int' or 'boolean' or 'buffer'
+        - 'size', either an integer of a string which is the 'name' of another
+          argument of type 'int'
+        - 'interface_index', for all `input` that should appear in JS, position
+          inside the argument list of the JS function
+        - 'tests', a list of values for this arguments, each value corresponding
+          to a different test case
+      - 'return', the return type of the WebAssembly function
+  */
+
+  const CheckIfByteArray = function(candidate, length, name) {
+    if (!(typeof(candidate) === "object") ||
+      !(candidate.length === length) ||
+      !(candidate.constructor === Uint8Array)
+    ) {
+      throw new Error(
+        "name: Please ensure the argument " + candidate + " is a " + length + "-bytes Uint8Array."
+      );
+    }
+  };
+
+  const malloc_array = function(array) {
```

Comment:
seems more like a stack allocation to me? maybe `alloca_array`?

Reply:
or even `copy_array_to_stack`

Reply:
Done